File: Gaussian Elimination.py
def rref(A):
# RREF  This function performs Gaussian Elimination using pivoting to convert a matrix A into I.
# 
#   Input:
#       A - A m*n dimensional matrix.

#   Output
#       R - An indentity matrix of dimension m*n.
 
#   Initializing the indexing variables.
    R = A
    [i,j] = [0, 0]
    [m, n] = [len(A), len(A[0])]

    while i<m and j<n:
#       PIVOTING
#       Finding the largest value in column j below row i.
        kmax = R[i][j]
        for k in range(i,m):
            if R[k][j]>kmax:
                kmax = R[k][j]
                knew = k
            else:
                continue
#       Swapping row k and row i. 
        for a in range(0,n):
            R[i][a], R[k][a] = R[k][a], R[i][a]

#       RREF MAIN STEPS
        if abs(R[i][j]) > pow(10,-16):
#           Normalizing row i.
            a = R[i][j]
            for k in range(j,n):
                R[i][k] /= a 
#           Eliminating values in column j other than row i.
            for k in range(0,m):
                a = R[k][j]
                if(k==i):
                    continue
                else:
                    for l in range(j,n):
                        R[k][l] -= a*R[i][l]
#       Preparing Next Steps
            i += 1
            j += 1
        else:
            j += 1
    
    return R

if "__main__":
    A = [[2, 1, -1], [3, -1, 2], [-2, 1, 2]]
    R = rref(A)
    print(R)

File: gaussian_elimination.m
function [R] = gaussian_elimination(A)
%GAUSSIAN_ELIMINATION  Returns Row Reduced Echelon Form of a matrix A.
%
%   Input:
%       A - Input Matrix. Should be a Matrix with valid values.
%
%   Output:
%       R - Row Reduced Echelon Form of A. Matrix with same dimension of A.

%   Base Case R and starting with row 1 and column 1.
    R = A;
    i=1; 
    j=1;
    [m,n] = size(R);

%   Operating on the new matrix
    while i<=m && j<=n
%       Finding Row with the largest absolute jth element.
        k = i;
        km = abs(R(i,j));
        for a = i:m
            if abs(R(a,j)) > km
                k = a;
                km = abs(R(a,j));
            end
        end

%       Swapping row i and k
        for b = 1:n
            temp = R(i,b);
            R(i,b) = R(k,b);
            R(k,b) = temp;
        end

%       Row Reduced Echelon Form Operations.        
        if km>10^-16
            a = R(i,j);
            for k = j:n
                R(i,k) = R(i,k)/a;
            end
            for k = [1:i-1, i+1:m]
                a = R(k,j);
                for l = j:n
                    R(k,l) = R(k,l) - a * R(i,l);
                end
            end
            i = i+1;
            j = j+1;
        else
            j = j+1;
        end
    end

end % umathur_pp6

